src/Kubernetes/Webhook.hs

```diff
@@ -0,0 +1,147 @@
+--
+-- @
+--   module Kubernetes.Example
+--       ( startApp,
+--         app,
+--       )
+--     where
+--
+--   import Control.Monad.IO.Class
+--   import qualified Data.Aeson as A
+--   import qualified Data.ByteString as BS
+--   import qualified Data.HashMap.Strict as HM
+--   import Data.Text
+--   import GHC.Generics
+--   import qualified Kubernetes.Webhook as W
+--   import Network.Wai
+--   import Network.Wai.Handler.Warp
+--   import Network.Wai.Handler.WarpTLS
+--   import Servant
+--   import System.Environment
+--
+--   type API =
+--     "mutate" :> ReqBody '[JSON] W.AdmissionReviewRequest :> Post '[JSON] W.AdmissionReviewResponse
+--
+--   data Toleration
+--     = Toleration
+--         { effect :: Maybe TolerationEffect,
+--           key :: Maybe Text,
+--           operator :: Maybe TolerationOperator,
+--           tolerationSeconds :: Maybe Integer,
+--           value :: Maybe Text
+--         }
+--     deriving (Generic, A.ToJSON)
+--
+--   data TolerationEffect = NoSchedule | PreferNoSchedule | NoExecute deriving (Generic, A.ToJSON)
+--
+--   data TolerationOperator = Exists | Equal deriving (Generic, A.ToJSON)
+--
+--   testToleration :: Toleration
+--   testToleration =
+--     Toleration
+--       { effect = Just NoSchedule,
+--         key = Just "dedicated",
+--         operator = Just Equal,
+--         tolerationSeconds = Nothing,
+--         value = Just "test"
+--       }
+--
+--   startApp :: IO ()
+--   startApp = do
+--     let tlsOpts = tlsSettings "/certs/tls.crt" "/certs/tls.key"
+--         warpOpts = setPort 8080 defaultSettings
+--     runTLS tlsOpts warpOpts app
+--
+--   app :: Application
+--   app = serve api server
+--
+--   api :: Proxy API
+--   api = Proxy
+--
+--   server :: Server API
+--   server = mutate
+--
+--   mutate :: W.AdmissionReviewRequest -> Handler W.AdmissionReviewResponse
+--   mutate req = pure $ W.mutatingWebhook req (\_ -> Right W.Allowed) addToleration
+--
+--   addToleration :: W.Patch
+--   addToleration =
+--     W.Patch
+--       [ W.PatchOperation
+--           { op = W.Add,
+--             path = "/spec/tolerations/-",
+--             from = Nothing,
+--             value = Just $ A.toJSON testToleration
+--           }
+--       ]
+-- @
+module Kubernetes.Webhook
+  ( mutatingWebhook,
+    validatingWebhook,
+    Allowed (..),
+    module Kubernetes.Webhook.Types,
+  )
+where
+
+import Data.Either
+import Kubernetes.Webhook.Types
+
+data Allowed = Allowed
+
+-- | Lets you create a mutating admission webhook
+mutatingWebhook ::
+  -- | the request the webhook receives from Kubernetes
+  AdmissionReviewRequest ->
+  -- | logic to validate the request or reject it with an error
+  (AdmissionRequest -> Either Status Allowed) ->
+  -- | the change to apply to the object
+  Patch ->
+  -- | the response sent back to Kubernetes
+  AdmissionReviewResponse
+mutatingWebhook AdmissionReviewRequest {request = req} allow patch =
```

Comment:
I realize I'm a little late to the party, but it seems that it rather than just a patch, it might be better to have a `(AdmissionRequest -> Patch)` as the final param - i.e. generate a patch based on the request.
Or possibly even have the second arg be `(AdmissionRequest -> Either Status Patch)`, but that might be complecting things too much.

Reply:
You are not late :) The version released is just the first one and improvements are definitely possible.

I thought about it and actually initially I designed the API to look like you described, but in the end I decided to go for a simpler signature for a few reasons:
- we are only really interested in the elements of the object more than the request itself, and JSONPatch is quite powerful in referring to the object already (see the "from" field http://jsonpatch.com/#operations ). I realize that it's not as powerful as having the request/object here, but i think it covers most cases
- for the second arg, I don't think this will bring any benefit, as you have access to the request at call time anyway
- if you really need to do more advanced stuff, you have full access to the types and you can modify them as you please (digression: perhaps the types are actually a bit too liberal, because invalid states are representable - e.g. having allowed:false and status Nothing - but at this stage i care more about ergonomics than perfection)

Perhaps a middle ground could be that we keep these functions and we add more advanced functions with a more complete signature, and then we refactor the current ones to use the complete ones. What do you think? If you want (and have time to) to contribute the feature as well please do :)